include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -179,16 +179,16 @@ __serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, const _I
         if constexpr (__can_use_ternary_op_v<decltype(__rng1[__rng1_idx]), decltype(__rng2[__rng2_idx])>)
         {
             // This implementation is required for performance optimization
-            __rng3[__rng3_idx] = (!__rng1_idx_less_n1 || __rng1_idx_less_n1 && __rng2_idx_less_n2 &&
-                                                             __comp(__rng2[__rng2_idx], __rng1[__rng1_idx]))
+            __rng3[__rng3_idx] = (!__rng1_idx_less_n1 || (__rng1_idx_less_n1 && __rng2_idx_less_n2 &&
+                                                          __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])))
                                      ? __rng2[__rng2_idx++]
                                      : __rng1[__rng1_idx++];
         }
         else
         {
             // TODO required to understand why the usual if-else is slower then ternary operator
             if (!__rng1_idx_less_n1 ||
-                __rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx]))
+                (__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])))
```

Comment:
Why the redundant parentheses were added?

Reply:
It silences this warning:

```
oneDPL/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:182:99: warning: '&&' within '||' [-Wlogical-op-parentheses]
```

The compiler is a bit paranoid - we know the priority of && and ||. Etherealness, I prefer not to introduce even such warnings because it is a library code...

Reply:
I also think this improves clarity for someone reading the code, but I know that may just be personal preference.

Reply:
> The compiler is a bit paranoid

Absolutely agree. It is the same  if the compiler would warn about  `a*b + 7` and force to write `(a*b) + 7`.
I would suggest suppressing the warning.

Reply:
I agree with Adam and this is why I always defer to the idea that explicit intention should be expressed by the programmer in any situation which is close to ambiguous.  
Even if the person reading the code knows the precedence, the extra parens confirm that the original code was written with this intention in mind (and it is not a bug lurking).  

In the code, its cost is just a couple of characters.  For the person writing the code, its cost the small inconvenience of another commit if we someday enforce reducing warnings to a minimum.

Reply:
Unfortunately, it's not the compiler being paranoid, it's users asking for this. I had the same sort of argument some time ago when worked on TBB, and I was on the position that programmers should know the precedence of `&&` to `||` (after all, noone would put parenthesis around multiplication to "disambiguate" precedence with addition. But it seems too many programmers do it wrong.

So if we want to enable -Wall or alike as part of our testing strategy, to avoid excessive warnings for users who use it in their projects (and it's quite common these days), we, as library developers, have to bite the bullet and do things we do not like.

Reply:
An alternative option that I want to mention would be to selectively suppress warnings we disagree with via compiler pragmas: https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas. My understanding is that via the push / pop mechanism we can target code blocks in our headers without affecting warnings in user code. I think TBB does this in some places.

I do not like this option though and think that we should directly fix these warnings as others suggest.

Reply:
Right, header-scope suppression with pragmas is an option which we can consider. Thanks @mmichel11 for mentioning this.

Reply:
It seems the majority prefers to keep paratheses. Let me close the discussion.

Reply:
The democracy is great thing :-)